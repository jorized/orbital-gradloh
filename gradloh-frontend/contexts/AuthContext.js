import React, { createContext, useState } from 'react';
import * as SecureStore from 'expo-secure-store';
const AuthContext = createContext(null);
const { Provider } = AuthContext;

const AuthProvider = ({ children }) => {


	const [authState, setAuthState] = useState({
		accessToken: null,
		refreshToken: null,
		authenticated: null
	});

	const logout = async () => {

		await SecureStore.deleteItemAsync('token');
		await SecureStore.deleteItemAsync('userdetails');
		await SecureStore.deleteItemAsync('selectedGradient');
		setAuthState({
			accessToken: null,
			refreshToken: null,
			authenticated: false,
			firstTimeUser: false
		});
	};

	const getAccessToken = () => {
		return authState.accessToken;
	};


	return (
		<Provider
			value={{
				authState,
				getAccessToken,
				setAuthState,
				logout,
			}}
		>
			{children}
		</Provider>
	);
};

export { AuthContext, AuthProvider };
