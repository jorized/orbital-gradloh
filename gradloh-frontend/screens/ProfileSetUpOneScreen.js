import {
	View,
	Text,
	StyleSheet,
	TextInput,
	Pressable,
	Platform,
	TouchableNativeFeedback,
	TouchableOpacity
} from 'react-native';
import { useFonts } from 'expo-font';
import {
	Lexend_300Light,
	Lexend_400Regular,
	Lexend_600SemiBold,
	Lexend_700Bold
} from '@expo-google-fonts/lexend';
import { Dropdown } from 'react-native-element-dropdown';
import { useEffect, useState } from 'react';
import { useNavigation, useRoute } from '@react-navigation/native';
import Icon from 'react-native-vector-icons/MaterialIcons';
import DropdownWithErrorHandling from '../components/DropdownWithErrorHandling';
import axios from 'axios';
import SelectableButton from '../components/SelectableButton';

export default function ProfileSetUpOneScreen() {
	const navigation = useNavigation();

	const [value, setValue] = useState('');
	const [isFocus, setIsFocus] = useState(false);
	const [error, setError] = useState(false);
	const [loading, setLoading] = useState(false);

	const route = useRoute();
	const { nickname, email, password } = route.params || {};

	const data = [
		{ label: '2022/2023', value: '2022-2023' },
		{ label: '2023/2024', value: '2023-2024' }
	];

	const [fontsLoaded] = useFonts({
		Lexend_300Light,
		Lexend_400Regular,
		Lexend_600SemiBold,
		Lexend_700Bold
	});

	const handleSubmit = () => {
		if (!value) {
			setError(true);
		} else {
			setError(false);
			// Handle the submission
			navigation.push('ProfileSetUpTwoScreen', {
				nickname: nickname,
				email: email,
				password: password,
				enrolmentYear: value
			});
		}
	};

	if (!fontsLoaded) {
		return <Text>Loading...</Text>;
	}

	return (
		<View style={styles.container}>
			<View style={styles.form}>
				<Text style={styles.profileSetUpTitle}>Profile Setup</Text>
				<Text style={styles.matricText}>
					*Please enter your matriculation year
				</Text>

				<DropdownWithErrorHandling
					data={data}
					value={value}
					placeholder="Select AY*"
					error={error}
					setValue={setValue}
					setIsFocus={setIsFocus}
					setError={setError}
				/>
			</View>
			{Platform.OS === 'android' ? (
				<TouchableNativeFeedback
					onPress={loading ? null : handleSubmit}
					background={TouchableNativeFeedback.Ripple('#fff', false)}
					disabled={loading}
				>
					<View
						style={[
							styles.loginPressable,
							loading && styles.disabledPressable
						]}
					>
						{loading ? (
							<ActivityIndicator size="small" color="#FFF" />
						) : (
							<Text style={styles.loginText}>Next</Text>
						)}
					</View>
				</TouchableNativeFeedback>
			) : (
				<TouchableOpacity
					style={[
						styles.loginPressable,
						loading && styles.disabledPressable
					]}
					onPress={loading ? null : handleSubmit}
					disabled={loading}
				>
					{loading ? (
						<ActivityIndicator size="small" color="#FFF" />
					) : (
						<Text style={styles.loginText}>Next</Text>
					)}
				</TouchableOpacity>
			)}
		</View>
	);
}

const styles = StyleSheet.create({
	container: {
		padding: 36,
		flex: 1,
		backgroundColor: 'white',
		justifyContent: 'space-between'
	},
	profileSetUpTitle: {
		fontFamily: 'Lexend_600SemiBold',
		fontSize: 40,
		marginBottom: 24
	},
	matricText: {
		fontFamily: 'Lexend_300Light',
		fontSize: 16,
		marginBottom: 24
	},
	dropdown: {
		borderBottomWidth: 1,
		borderBottomColor: '#CBCBCB'
	},
	dropdownPlaceholder: {
		color: '#CBCBCB'
	},
	dropdownError: {
		borderBottomColor: '#D00E17'
	},
	dropdownErrorText: {
		color: '#D00E17'
	},
	loginPressable: {
		padding: 16,
		backgroundColor: '#EF7C00',
		borderRadius: 12
	},
	loginText: {
		textAlign: 'center',
		color: 'white',
		fontSize: 18,
		fontFamily: 'Lexend_400Regular'
	},
	errorContainer: {
		flexDirection: 'row',
		alignItems: 'center',
		marginTop: 5
	},
	errorIcon: {
		marginRight: 5
	},
	errorMessage: {
		color: '#D00E17',
		fontSize: 12
	}
});
