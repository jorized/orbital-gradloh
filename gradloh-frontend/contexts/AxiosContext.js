import React, { createContext, useContext, useEffect } from 'react';
import axios from 'axios';
import { AuthContext } from './AuthContext';
import { LoadingContext } from './LoadingContext'; // Import the Loading Context
import createAuthRefreshInterceptor from 'axios-auth-refresh';
import * as SecureStore from 'expo-secure-store';

const AxiosContext = createContext();
const authApiUrl = process.env.EXPO_PUBLIC_API_AUTH_URL;
const publicApiUrl = process.env.EXPO_PUBLIC_API_URL;
const { Provider } = AxiosContext;

const AxiosProvider = ({ children }) => {
  const authContext = useContext(AuthContext);
  const { setIsLoading } = useContext(LoadingContext); // Use the Loading Context
  const requestQueue = [];
  let isRefreshing = false;

  const processQueue = (error, token = null) => {
    requestQueue.forEach(promise => {
      if (error) {
        promise.reject(error);
      } else {
        promise.resolve(token);
      }
    });
    requestQueue.length = 0;
  };

      //Those who are not, use this route
    const publicAxios = axios.create({
      baseURL: authApiUrl
    });

  const authAxios = axios.create({
    baseURL: publicApiUrl,
    headers: {
      Authorization: `Bearer ${authContext.getAccessToken()}`,
    },
  });

  authAxios.interceptors.response.use(null, async (error) => {
    const originalRequest = error.config;
    if (error.response.status === 403 && !originalRequest._retry) {
      originalRequest._retry = true;
      if (!isRefreshing) {
        isRefreshing = true;
        setIsLoading(true); // Show loading indicator

        const data = {
          refreshToken: authContext.authState.refreshToken,
        };
        const options = {
          method: 'POST',
          data,
          url: `${authApiUrl}/refresh`,
        };

        try {
          const tokenRefreshResponse = await axios(options);
          isRefreshing = false;
          setIsLoading(false); // Hide loading indicator

          authContext.setAuthState({
            ...authContext.authState,
            accessToken: tokenRefreshResponse.data.accessToken,
          });

          await SecureStore.setItemAsync(
            'token',
            JSON.stringify({
              accessToken: tokenRefreshResponse.data.accessToken,
              refreshToken: authContext.authState.refreshToken,
            })
          );

          processQueue(null, tokenRefreshResponse.data.accessToken);

          originalRequest.headers.Authorization =
            'Bearer ' + tokenRefreshResponse.data.accessToken;

          return axios(originalRequest);
        } catch (e) {
          isRefreshing = false;
          setIsLoading(false); // Hide loading indicator
          processQueue(e, null);
          throw e;
        }
      }

      const retryOriginalRequest = new Promise((resolve, reject) => {
        requestQueue.push({ resolve, reject });
      });

      return retryOriginalRequest
        .then(token => {
          originalRequest.headers.Authorization = 'Bearer ' + token;
          return axios(originalRequest);
        })
        .catch(err => {
          return Promise.reject(err);
        });
    }
    return Promise.reject(error);
  });

  return (
    <Provider value={{ authAxios, publicAxios }}>
      {children}
    </Provider>
  );
};

export { AxiosContext, AxiosProvider };
