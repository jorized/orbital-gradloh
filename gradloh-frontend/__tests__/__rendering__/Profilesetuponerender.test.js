import React from 'react';
import TestRenderer, { act } from 'react-test-renderer';
import { NavigationContainer } from '@react-navigation/native';
import ProfileSetUpOneScreen from '../../screens/Auth/ProfileSetUpOneScreen';
import { AuthContext } from '../../contexts/AuthContext';
import { AxiosContext } from '../../contexts/AxiosContext';

jest.mock('react-native-toast-message', () => {
  const React = require('react');
  const Toast = React.forwardRef((props, ref) => <div ref={ref} {...props} />);
  return {
    show: jest.fn(),
    hide: jest.fn(),
    default: Toast,
  };
});

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useRoute: jest.fn().mockReturnValue({
      params: {
        nickname: 'test',
        email: 'test@u.nus.edu',
        password: 'Password123%',
      },
    }),
  };
});

jest.mock('expo-font', () => ({
  useFonts: () => [true],
}));

jest.mock('../../components/Auth/DropdownWithErrorHandling', () => 'DropdownWithErrorHandling');

const mockAuthContextValue = {
  authState: { authenticated: false },
  setAuthState: jest.fn(),
  getAccessToken: jest.fn(),
  logout: jest.fn(),
};

const mockAxiosContextValue = {
  publicAxios: {
    post: jest.fn().mockResolvedValue({ data: { message: 'Success' } }),
  },
  authAxios: {
    post: jest.fn().mockResolvedValue({ data: { message: 'Success' } }),
  },
};

describe('ProfileSetUpOneScreen', () => {
  it('renders correctly', async () => {
    let tree;

    await act(async () => {
      tree = TestRenderer.create(
        <AuthContext.Provider value={mockAuthContextValue}>
          <AxiosContext.Provider value={mockAxiosContextValue}>
            <NavigationContainer>
              <ProfileSetUpOneScreen />
            </NavigationContainer>
          </AxiosContext.Provider>
        </AuthContext.Provider>
      ).toJSON();
    });

    expect(tree).toMatchSnapshot();
  });
});
