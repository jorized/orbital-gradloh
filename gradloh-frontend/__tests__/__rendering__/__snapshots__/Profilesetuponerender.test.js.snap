// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileSetUpOneScreen renders correctly 1`] = `null`;
